stage #2.5/module #3 Spring Core/#5.1. Annotation-based configuration/content.md

```diff
@@ -0,0 +1,470 @@
+
+Even typed `Map` instances can be autowired as long as the expected key type is String. The map values contain all beans
+of the expected type, and the keys contain the corresponding bean names, as the following example shows:
+
+```java
+public class MusicRatingService {
+
+    private Map<String, MusicMagazine> musicMagazines;
+
+    @Autowired
+    public void setMusicMagazines(Map<String, MusicMagazine> musicMagazines) {
+        this.musicMagazines = musicMagazines;
+    }
+
+    // ...
+}
+```
+
+By default, autowiring fails when no matching candidate beans are available for a given injection point. In the case of
+a declared array, collection, or map, at least one matching element is expected.
+
+The default behavior is to treat annotated methods and fields as indicating required dependencies. You can change this
+behavior, enabling the framework to skip a non-satisfiable injection point through marking it as non-required (i.e., by
+setting the required attribute in `@Autowired` to false):
+
+```
+@Autowired(required = false)
+```
+
+A non-required method will not be called at all if its dependency (or one of its dependencies, in case of multiple
+arguments) is not available. A non-required field will not get populated at all in such cases, leaving its default value
+in place.
+
+Alternatively, you can express the non-required nature of a particular dependency through Java 8's `java.util.Optional`,
+as the following example shows:
+
+```java
+public class MusicRatingService {
+
+    private Optional<MusicRatesFinder> musicRatesFinder;
+
+    @Autowired
+    public void setMusicRatesFinder(Optional<MusicRatesFinder> musicRatesFinder) {
+        this.musicRatesFinder = musicRatesFinder;
+    }
+
+    // ...
+}
+```
+
+As of Spring Framework 5.0, you can also use a `@Nullable annotation` (of any kind in any package - for example,
+`javax.annotation.Nullable` from JSR-305)
+
+```java
+public class MusicRatingService {
+
+    private MusicRatesFinder musicRatesFinder;
+
+    @Autowired
+    public void setMusicRatesFinder(@Nullable MusicRatesFinder musicRatesFinder) {
+        this.musicRatesFinder = musicRatesFinder;
+    }
+
+    // ...
+}
+```
+
+You can also use `@Autowired` for interfaces that are well-known resolvable dependencies: `BeanFactory`,
+`ApplicationContext`, `Environment`, `ResourceLoader`, `ApplicationEventPublisher`, and `MessageSource`. These
+interfaces and their extended interfaces, such as `ConfigurableApplicationContext` or `ResourcePatternResolver`, are
+automatically resolved, with no special setup necessary.
+
+> The `@Autowired`, `@Inject`, `@Value`, and `@Resource` annotations are handled by Spring `BeanPostProcessor`
+> implementations. This means that you cannot apply these annotations within your own `BeanPostProcessor` or
+> `BeanFactoryPostProcessor` types (if any). These types must be 'wired up' explicitly by using XML or a Spring
+> `@Bean` method.
```

Comment:
Propably this paragraph is too difficult for students who don't know Spring at all

Reply:
Do you think this should be removed? I agree that students may not know internal structure of Spring and in our sections we don't describe that but the content above is marked as "note" (not main text), but such words as "BeanPostProcessor" or "BeanFactoryPostProcessor" could force students to dive deeper and learn more about Spring structure on their own.

Reply:
Ok, lets leave it